contracts/utils/Swap.sol

```diff
@@ -0,0 +1,504 @@
+     * @param minPrice minimum price of oToken denominated in biddingToken
+     * @param minBidSize minimum amount of oToken requested in a single bid
+     * @param totalSize amount of oToken offered by seller
+     */
+    function createOffer(
+        address oToken,
+        address biddingToken,
+        uint96 minPrice,
+        uint96 minBidSize,
+        uint128 totalSize
+    ) external override returns (uint256 swapId) {
+        require(oToken != address(0), "oToken cannot be the zero address");
+        require(
+            biddingToken != address(0),
+            "BiddingToken cannot be the zero address"
+        );
+        require(minPrice > 0, "MinPrice must be larger than zero");
+        require(minBidSize > 0, "MinBidSize must be larger than zero");
+        require(minBidSize <= totalSize, "MinBidSize exceeds total size");
+
+        offersCounter += 1;
+
+        swapId = offersCounter;
+
+        swapOffers[swapId].seller = msg.sender;
+        swapOffers[swapId].oToken = oToken;
+        swapOffers[swapId].biddingToken = biddingToken;
+        swapOffers[swapId].minBidSize = minBidSize;
+        swapOffers[swapId].minPrice = minPrice;
+        swapOffers[swapId].totalSize = totalSize;
+        swapOffers[swapId].availableSize = totalSize;
+        // We warm the storage slot with 1 wei so we avoid a cold SSTORE
+        swapOffers[swapId].totalSales = 1;
+
+        emit NewOffer(
+            swapId,
+            msg.sender,
+            oToken,
+            biddingToken,
+            minPrice,
+            minBidSize,
+            totalSize
+        );
+    }
+
+    /**
+     * @notice Settles the swap offering by iterating through the bids
+     * @param swapId unique identifier of the swap offer
+     * @param bids bids for swaps
+     */
+    function settleOffer(uint256 swapId, Bid[] calldata bids)
+        external
+        override
+        nonReentrant
+    {
+        Offer storage offer = swapOffers[swapId];
+
+        address seller = offer.seller;
+        uint256 availableSizeBefore = offer.availableSize;
+        require(
+            seller == msg.sender,
+            "Only seller can settle or offer doesn't exist"
+        );
+        require(availableSizeBefore > 0, "Offer fully settled");
+
+        uint256 totalSales;
+        OfferDetails memory offerDetails;
+        offerDetails.seller = seller;
+        offerDetails.oToken = offer.oToken;
+        offerDetails.biddingToken = offer.biddingToken;
+        offerDetails.minPrice = offer.minPrice;
+        offerDetails.minBidSize = offer.minBidSize;
+
+        for (uint256 i = 0; i < bids.length; i++) {
+            _swap(offerDetails, offer, bids[i]);
+            totalSales += bids[i].sellAmount;
+        }
+
+        uint256 availableSizeAfter = offer.availableSize;
+        // Deduct the initial 1 wei offset if offer is fully settled
+        offer.totalSales += totalSales - (availableSizeAfter == 0 ? 1 : 0);
```

Comment:
Remove condition `offer.availableSize == 0` before updating `offer.totalSales`. This is to account for partial settlement, otherwise `offer.totalSales` will remain 0 until offer is fully settled, causing `averagePriceForOffer` to return 0 as well.